Cleanup error handling conditions for ACA compute environment

eerhardt opened the pull request:
## Description

1. When calling PublishAsAzureContainerApp and there are no ACA compute environments, throw. Also throw if the environment is wrong.
2. Support multiple ACA environments

Contributes to #9532

Will do the other 3 compute environments separately after we agree upon a pattern.

## Checklist

- Is this feature complete?
  - [x] No. Follow-up changes expected. **Need to do the same for the other 3 compute environments**
- Are you including unit tests for the changes and scenario tests if relevant?
  - [x] Yes
- Did you add public API?
  - [x] Yes
    - If yes, did you have an API Review for it?
      - [x] No
    - Did you add `<remarks />` and `<code />` elements on your triple slash comments?
      - [x] No
- Does the change make any security assumptions or guarantees?
  - [x] No
- Does the change require an update in our Aspire docs?
  - [x] No

eerhardt committed:
Cleanup error handling conditions for ACA compute environment

1. When calling PublishAsAzureContainerApp and there are no ACA compute environments, throw. Also throw if the environment is wrong.
2. Support multiple ACA environments

Contributes to #9532

eerhardt commented on src/Aspire.Hosting.Azure.AppContainers/AzureContainerAppsInfrastructure.cs:
```diff
@@ -27,51 +27,48 @@ public async Task BeforeStartAsync(DistributedApplicationModel appModel, Cancell
+                environment);
 
-        var environment = caes.FirstOrDefault();
+            foreach (var r in appModel.GetComputeResources())
```
This won't work because this is the method addign the DeploymentTargetAnnotation.

eerhardt committed:
Fix tests
Fix spelling

eerhardt commented on src/Aspire.Hosting.Azure.AppContainers/AzureContainerAppsInfrastructure.cs:
```diff
@@ -27,51 +27,48 @@ public async Task BeforeStartAsync(DistributedApplicationModel appModel, Cancell
+    }
 
-        foreach (var r in appModel.Resources)
+    private static void EnsurePublishAsAcaAnnotationsMatch(DistributedApplicationModel appModel, bool hasAcaEnvironments)
```
1. How is this too early? We've added all the ACA deployment targets we are going to add by this point.
2.
> Let it fail at the normal place during publish.

How would this fail at the "normal place during publish"? If you set some other compute environment, or didn't add an ACA environment, it isn't going to fail.

davidfowl replied:
I can call `WithComputeEnvironment`  or add the ComputeEnvironmentAnnotation in BeforeStart and this shouldn't fail. It should fail when somebody queries the deployment target at publish time and it's ambiguous (just like it happens today when you use different types of compute environments).

eerhardt replied:
OK, I changed this logic to only throw when you have `AzureContainerAppCustomizationAnnotation` and no ACA environments.

captainsafia commented on src/Aspire.Hosting.Azure.AppContainers/AzureContainerAppsInfrastructure.cs:
```diff
@@ -27,51 +27,48 @@ public async Task BeforeStartAsync(DistributedApplicationModel appModel, Cancell
         var caes = appModel.Resources.OfType<AzureContainerAppEnvironmentResource>().ToArray();
 
-        if (caes.Length > 1)
+        foreach (var environment in caes)
```
It seems like this code block is unconditionally associating each compute resource with every compute environment in the application. Don't we want to respect the `ComputeEnvironmentAnnotation` here?

Also, I think we'll need some sort of exception handling that ensures that if there are multiple environments registered then `WithComputeEnvironment` must be called? We might already have that implemented but not sure...

davidfowl replied:
That happens later at query time.

eerhardt replied:
> It seems like this code block is unconditionally associating each compute resource with every compute environment in the application. Don't we want to respect the ComputeEnvironmentAnnotation here?

This was the agreed upon design. Every compute environment adds its DeploymentTarget, and then the correct one gets selected out later here:

https://github.com/dotnet/aspire/blob/c5afd7150a85acd217ff79f9d41d459a0914f76f/src/Aspire.Hosting/ApplicationModel/ResourceExtensions.cs#L579-L613

When you call that method without specifying a compute environment, it throws if there are multiple and you didn't specify a `ComputeEnvironmentAnnotation`.

It gets called in a few spots:

https://github.com/dotnet/aspire/blob/c5afd7150a85acd217ff79f9d41d459a0914f76f/src/Aspire.Hosting/Publishing/ManifestPublishingContext.cs#L158

https://github.com/dotnet/aspire/blob/c5afd7150a85acd217ff79f9d41d459a0914f76f/src/Aspire.Hosting/Publishing/ManifestPublishingContext.cs#L254

https://github.com/dotnet/aspire/blob/c5afd7150a85acd217ff79f9d41d459a0914f76f/src/Aspire.Hosting.Azure/AzurePublishingContext.cs#L233

eerhardt committed:
PR feedback

Only throw if there are no ACA environments. Other misconfigurations are not checked.
